### PR TITLE
feat: manually cascade soft deletes for users relations

### PR DESCRIPTION
**Pull request by jpmcb**
## Description

TypeORM doesn't seem to play well with our use of foreign key relations during soft deletes. I tried a number of configurations proposed in https://github.com/typeorm/typeorm/issues/1460 and https://github.com/typeorm/typeorm/issues/5877 - I consistently got errors where the `user_id` was attempted to be updated to `NULL` which violated our foreign key constraints.

Instead, this manually gets the user's relations and manually soft deletes them.

Tested on my test user:

Correctly deletes user:
```
id         deleted_at                  login
137924913  2023-10-27 19:17:09.860376  not-linus-torvalds	
```

Correctly deletes highlight:
```
id   user_id    deleted_at
245  137924913  2023-10-27 14:57:26.229
```

Correctly deletes insight page:
```
id   user_id    deleted_at
307  137924913  2023-10-27 19:17:09.969833
```

Correctly deletes list:
```
id                                    user_id    deleted_at
5e848867-5ec4-4520-8796-e98af0885cce  137924913  2023-10-27 19:17:10.004279
```

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Followup to #395 

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

Added some inline docs to keep this thread alive

### CODE REVIEW

File: src/user/services/user.service.ts [hunk 1]
```diff
@@ -374,6 +405,41 @@ export class UserService {
...
+          await this.userCollaborationRepository.softDelete(req_collab.id);
+        })
+      );
+
+      // soft delete the user's lists
+      await Promise.all(
+        userAndRelations.lists.map(async (list) => {
+          await this.userListRepository.softDelete(list.id);
+        })
+      );
```
Comment (nickytonline) [inline src/user/services/user.service.ts:441]: question: can these all be done in parallel, or do the `Promise.all`s need to waterfall?

**Comment by jpmcb** [comment]
Force pushed to address failing tests 📝

File: src/user/services/user.service.ts (hunk 1 above)
Comment (jpmcb) [reply src/user/services/user.service.ts:441]: They can probably all be done in parallel 👍🏼 good suggestion!

**Comment by jpmcb** [comment]
Force pushed to make the waterfall `Promise.all` happen all in parallel - thanks @nickytonline ! 💨 🚀

File: src/user/services/user.service.ts [hunk 2]
```diff
@@ -374,8 +405,45 @@ export class UserService {
...
+        Promise.all(
+          userAndRelations.lists.map(async (list) => {
+            await this.userListRepository.softDelete(list.id);
+          })
+        ),
+      ]);
+
       return user;
     } catch (e) {
+      console.log(e);
```
Comment (jpmcb) [inline src/user/services/user.service.ts:446]: 🙈

File: src/insight/entities/insight.entity.ts [hunk 3]
```diff
@@ -115,7 +115,7 @@ export class DbInsight extends BaseEntity {
   public deleted_at?: Date;
 
   @ApiHideProperty()
-  @ManyToOne(() => DbUser, (user) => user.insights)
+  @ManyToOne(() => DbUser, (user) => user.insights, { onDelete: "CASCADE" })
```
Comment (brandonroberts) [inline src/insight/entities/insight.entity.ts:118]: Question: We still need the `onDelete` even though we're doing the softRemove calls manually?

Comment (jpmcb) [reply src/insight/entities/insight.entity.ts:118]: We don't need them for soft deletes. But I added them so we'll have them when real deletes happen and to match the cascades we have in some of the database table definitions

Comment (brandonroberts) [reply src/insight/entities/insight.entity.ts:118]: Got it

**Review by brandonroberts — APPROVED** [review]

File: src/user/services/user.service.ts [hunk 4]
```diff
@@ -374,6 +405,42 @@ export class UserService {
         is_open_sauced_member: false,
       });
 
+      await Promise.all([
```
Comment (nickytonline) [inline src/user/services/user.service.ts:408]: This should be good for now but depending on how much content a user has, [p-limit](https://www.npmjs.com/package/p-limit) is probably worth looking into in the future.

**Review by nickytonline — APPROVED** [review]
Thanks for making the change @jpmcb!  :shipit:

**Pull request merged**